File: src/main/java/ExistingIntermediaryNamesProvider.java

import com.intellij.openapi.components.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;


// this doesn't work in kotlin for some fucking reason

// So we can find methods that have intermediary descriptors while the user sees named descriptors
@State(name = "RenamedNames", storages = @Storage(StoragePathMacros.CACHE_FILE))
public class ExistingIntermediaryNamesProvider implements PersistentStateComponent<ExistingIntermediaryNamesProvider.State> {

    public static class State {
        public List<String> existingIntermediaryNames = new ArrayList<>();
        @Nullable
        public String currentVersion = null;

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            State state = (State) o;
            return Objects.equals(existingIntermediaryNames, state.existingIntermediaryNames);
        }

        @Override
        public int hashCode() {
            return Objects.hash(existingIntermediaryNames);
        }
    }

    private State state = new State();

    /**
     * Must be called whenever switching to a new version
     */
    public void update(String newVersion) {
        
    }

    public boolean intermediaryNameExists(String name,String version) {
        if(state.existingIntermediaryNames.isEmpty() || !Objects.equals(state.currentVersion, version)) update(version);
        return state.existingIntermediaryNames.contains(name);
    }

    public State getState() {
        return this.state;
    }

    public void loadState(@NotNull State state) {
        this.state = state;
    }

    public static RenamedNamesProvider getInstance() {
        return ServiceManager.getService(RenamedNamesProvider.class);
    }

}

File: src/main/kotlin/cloak/idea/ExistingIntermediaryNamesProvider.kt
package cloak.idea

import cloak.fabric.Intermediary.fetchExistingNames
import cloak.util.IdeaCache
import cloak.util.IdeaCacheGetter
import com.intellij.openapi.components.State
import com.intellij.openapi.components.Storage
import com.intellij.openapi.components.StoragePathMacros
import kotlinx.serialization.Serializable

@Serializable
data class LatestIntermediaryNames(
    val classNames: Set<String> = HashSet(),
    /** Map from intermediary name to descriptor*/
    val methodNames: Map<String, String> = HashMap(),
    val fieldNames: Map<String, String> = HashMap(),
    var currentVersion: String? = null
)


@State(
    name = "ExistingIntermediaryNames",
    storages = [Storage(StoragePathMacros.CACHE_FILE)]
)
class ExistingIntermediaryNamesProvider :
    IdeaCache<LatestIntermediaryNames>(LatestIntermediaryNames()) {

    //TODO: adjust to be able to fetch descriptors and stuff
    /**
     * Must be called whenever switching to a new version
     */
    fun update(newVersion: String) {
        println("Updating intermediary names")
        state = fetchExistingNames(newVersion)
        state.currentVersion = newVersion
    }

    fun getNames(version: String): LatestIntermediaryNames {
        if (state.currentVersion != version) update(version)
        return state
    }

    companion object : IdeaCacheGetter<ExistingIntermediaryNamesProvider>(ExistingIntermediaryNamesProvider::class.java)

}